pkg/agent/controller/ipseccertificate/ipsec_certificate_controller.go

```diff
@@ -0,0 +1,520 @@
+		return time.Now()
+	}
+	notAfter := pair.certificate[0].NotAfter
+	totalDuration := float64(notAfter.Sub(pair.certificate[0].NotBefore))
+	deadline := pair.certificate[0].NotBefore.Add(jitteryDuration(totalDuration))
+	klog.Infof("Certificate expiration is %v, rotation deadline is %v", notAfter, deadline)
+	return deadline
+}
+
+func (c *Controller) syncConfigurations() error {
+	startTime := time.Now()
+	defer func() {
+		d := time.Since(startTime)
+		klog.V(2).Infof("Finished syncing certificate configurations (%v)", d)
+	}()
+	caData, ca, err := loadRootCA(c.caPath)
+	if err != nil {
+		return err
+	}
+	keyData, key, err := loadOrCreatePrivateKey(c.privateKeyPath)
+	if err != nil {
+		return err
+	}
+	certData, cert, err := loadCertificate(c.certPath)
+	if err != nil {
+		return err
+	}
+	if _, ok := key.(crypto.Signer); !ok {
+		return fmt.Errorf("error reading key: key did not implement crypto.Signer")
+	}
+	newPair := &certificateKeyPair{
+		rawCert:       certData,
+		rawKey:        keyData,
+		rawCACert:     caData,
+		caCertificate: ca,
+		certificate:   cert,
+		privateKey:    key.(crypto.Signer),
+	}
+	c.certificateKeyPair.Store(newPair)
+	err = newPair.validate()
+	if err != nil {
+		c.renewEventChan <- struct{}{}
+		klog.ErrorS(err, "Current certificate configurations are not valid")
+		return nil
+	}
+	return c.configureOVS()
+}
+
+func loadRootCA(caPath string) ([]byte, []*x509.Certificate, error) {
+	pemBlock, err := ioutil.ReadFile(caPath)
+	if err != nil {
+		return nil, nil, err
+	}
+	certs, err := certutil.ParseCertsPEM(pemBlock)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error reading %s: %s", caPath, err)
+	}
+	return pemBlock, certs, nil
+}
+
+func loadOrCreatePrivateKey(privateKeyPath string) ([]byte, crypto.PrivateKey, error) {
+	var keyPEMBytes []byte
+	_, err := os.Stat(privateKeyPath)
+	if err == nil {
+		// load the private key contents from file
+		keyPEMBytes, err = ioutil.ReadFile(privateKeyPath)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to read key file %s: %v", privateKeyPath, err)
+		}
+	} else if err != nil && !os.IsNotExist(err) {
+		return nil, nil, fmt.Errorf("failed to stat key file %s: %v", privateKeyPath, err)
+	}
+	if len(keyPEMBytes) > 0 {
+		// try to parse private key from existing file
+		privateKey, err := keyutil.ParsePrivateKeyPEM(keyPEMBytes)
+		if err != nil {
+			klog.ErrorS(err, "Parse key from file error", "file", privateKeyPath)
+		} else {
+			return keyPEMBytes, privateKey, nil
+		}
+	}
+	// The key file does not exist or has been deleted. Generate a new key.
+	klog.Info("Generating new private key for IPSec")
+	key, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to generate new RSA private key: %v", err)
+	}
+	desiredKeyBytes, err := keyutil.MarshalPrivateKeyToPEM(key)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to marshal privatekey: %v", err)
+	}
+	if bytes.Equal(desiredKeyBytes, keyPEMBytes) {
```

Comment:
At this moment, keyPEMBytes should be empty or invalid while desiredKeyBytes should be a random key, how could they equal?

Reply:
Thanks for pointing it out. This logic was used to sync known key data to files when there is a file watcher. This should be removed.